apps/omg_watcher/lib/omg_watcher/db/eth_event.ex

```diff
@@ -72,44 +77,50 @@ defmodule OMG.Watcher.DB.EthEvent do
   @spec insert_exits!([non_neg_integer()]) :: :ok
   def insert_exits!(exits) do
     exits
-    |> Stream.map(&utxo_pos_from_exit_event/1)
+    |> Stream.map(&utxo_exit_from_exit_event/1)
     |> Enum.each(&insert_exit!/1)
   end
 
-  @spec utxo_pos_from_exit_event(%{call_data: %{utxo_pos: pos_integer()}}) :: Utxo.Position.t()
-  defp utxo_pos_from_exit_event(%{call_data: %{utxo_pos: utxo_pos}}), do: Utxo.Position.decode!(utxo_pos)
+  @spec utxo_exit_from_exit_event(%{call_data: %{utxo_pos: pos_integer()}, rootchain_txhash: charlist()}) :: Utxo.Exit.t()
+  defp utxo_exit_from_exit_event(%{call_data: %{utxo_pos: utxo_pos}, rootchain_txhash: rootchain_txhash}) do
+    %{rootchain_txhash: rootchain_txhash, decoded_utxo_position: Utxo.Position.decode!(utxo_pos)}
+  end
+
+  @spec insert_exit!(Utxo.Exit.t()) :: {:ok, %__MODULE__{}} | {:error, Ecto.Changeset.t()}
+  defp insert_exit!(%{rootchain_txhash: rootchain_txhash, decoded_utxo_position: decoded_utxo_position}) do\
+    case get(rootchain_txhash) do
+      nil ->
+        ethevent = %__MODULE__{
+          rootchain_txhash: rootchain_txhash,
+          event_type: :standard_exit,
+        }
 
-  @spec insert_exit!(Utxo.Position.t()) :: {:ok, %__MODULE__{}} | {:error, Ecto.Changeset.t()}
-  defp insert_exit!(Utxo.position(blknum, txindex, _oindex) = position) do
-    {:ok, _} =
-      if existing_exit = get(exit_key(position)) != nil do
+        DB.TxOutput.get_by_position(decoded_utxo_position)
+        |> txoutput_changeset(%{childchain_utxohash: generate_childchain_utxohash(decoded_utxo_position)}, ethevent)
+        |> DB.Repo.update!()
+
+      existing_exit ->
         {:ok, existing_exit}
-      else
-        utxo = DB.TxOutput.get_by_position(position)
+    end
+  end
 
-        %__MODULE__{
-          hash: exit_key(position),
-          blknum: blknum,
-          txindex: txindex,
-          event_type: :exit,
-          exited_utxo: utxo
-        }
-        |> DB.Repo.insert()
-      end
+  def txoutput_changeset(struct, params, ethevent) do
+    fields = [:blknum, :txindex, :oindex, :owner, :amount, :currency, :childchain_utxohash]
+
+    struct
+    |> DB.Repo.preload(:ethevents)
+    |> cast(params, fields)
+    |> put_assoc(:ethevents, [ethevent])
+    |> validate_required(fields)
   end
 
   @doc """
-  Good candidate for deposit/exit primary key is a pair (Utxo.position, event_type).
-  Switching to composite key requires careful consideration of data types and schema change,
-  so for now, we'd go with artificial key
+  Generate a unique childchain_utxohash from the Utxo.position
   """
-  @spec generate_unique_key(Utxo.Position.t(), :deposit | :exit) :: OMG.Crypto.hash_t()
-  def generate_unique_key(position, type) do
-    "<#{position |> Utxo.Position.encode()}:#{type}>" |> Crypto.hash()
+  @spec generate_childchain_utxohash(Utxo.Position.t()) :: OMG.Crypto.hash_t()
```

Comment:
Ouch, I'm scared here... can you elaborate why it will be unique for such events:

* deposited funds to `position = {1, 0, 0}` 
* exited funds from `position = {1, 0, 0}`
Consider adding comment

Reply:
Maybe it's just the function name that is confusing? I can rename it if it makes things clearer.

Just for clarity here is the many-to-many:

```
txoutputs                        ethevents_txoutputs                        ethevents
{1, 0, 0}        <------>        0x123, {1, 0, 0}          <------->        0x123   deposit
                                 0x234, {1, 0, 0}          <------->        0x234   exit
```

We see it's the same utxo, but with several ethevents.

`{1, 0, 0}` is the composite primary key into `txoutputs`. So whatever hash we generate from the position value will also be unique in the `txoutputs` table. And every `{position, eth_txn}` will also be unique in the `ethevents_txoutputs` table.

(As a side note, you may ask why I created a new childchain_utxohash column which is just a 'fake' primary-key column, instead of just using the composite key directly in the `ethevents_txoutputs` table. I did try to do this but ecto sadly doesn't seem to support composite keys in many-to-many mappings 😞 . The problem is here: https://github.com/omisego/elixir-omg/blob/003c36c8fccc0ee915b3ff6677aed407495bf351/apps/omg_watcher/lib/omg_watcher/db/eth_event.ex#L37  The `join_keys`, accepts only one column per side of the association.)

Reply:
> but ecto sadly doesn't seem to support composite keys in many-to-many mappings disappointed

Yes I was hit by this in my optimization efforts